Userland/Libraries/LibWeb/WebAudio/BaseAudioContext.cpp

```diff
@@ -120,4 +124,122 @@ void BaseAudioContext::queue_a_media_element_task(Function<void()> steps)
     HTML::main_thread_event_loop().task_queue().add(move(task));
 }
 
+// https://webaudio.github.io/web-audio-api/#dom-baseaudiocontext-decodeaudiodata
+JS::NonnullGCPtr<JS::Promise> BaseAudioContext::decode_audio_data(JS::Handle<WebIDL::BufferSource> audio_data, JS::GCPtr<WebIDL::CallbackType> success_callback, JS::GCPtr<WebIDL::CallbackType> error_callback)
+{
+    auto& realm = this->realm();
+
+    // FIXME: When decodeAudioData is called, the following steps MUST be performed on the control thread:
+
+    // 1. If this's relevant global object's associated Document is not fully active then return a
+    //    promise rejected with "InvalidStateError" DOMException.
+    auto const& associated_document = verify_cast<HTML::Window>(HTML::relevant_global_object(*this)).associated_document();
+    if (!associated_document.is_fully_active()) {
+        auto error = WebIDL::InvalidStateError::create(realm, "The document is not fully active."_string);
+        return WebIDL::create_rejected_promise_from_exception(realm, error);
+    }
+
+    // 2. Let promise be a new Promise.
+    auto promise = WebIDL::create_promise(realm);
+
+    // FIXME: 3. If audioData is detached, execute the following steps:
+    if (true) {
+        // FIXME: 3.1. Append promise to [[pending promises]].
+
+        // FIXME: 3.2. Detach the audioData ArrayBuffer. If this operations throws, jump to the step 3.
+
+        // 3.3. Queue a decoding operation to be performed on another thread.
+        queue_a_decoding_operation(promise, move(audio_data), success_callback, error_callback);
+    }
+
+    // 4. Else, execute the following error steps:
+    else {
+        // 4.1. Let error be a DataCloneError.
+        auto error = WebIDL::DataCloneError::create(realm, "Audio data is not detached."_string);
+
+        // 4.2. Reject promise with error, and remove it from [[pending promises]].
+        WebIDL::reject_promise(realm, promise, error);
+
+        // 4.3. Queue a media element task to invoke errorCallback with error.
+        if (error_callback) {
+            queue_a_media_element_task([&] {
```

Comment:
I am not sure if it is safe for us to capture by reference the error callback / allows the GC to see that properly. Generally I think we explicitly capture GCPtr's by value. Also makes me think we should make `queue_a_media_element_task` take a `HeapFunction` directly because it can end up simplifying memory management in the future (for example, like fc83653f3c7a650229d2be3eef0f1738a9395db9)

Reply:
Made it accept a `JS::HeapFunction` and captured the callbacks by value. Thanks!